src/plugins/intel_gpu/include/intel_gpu/runtime/lru_cache.hpp

```diff
@@ -164,6 +179,49 @@ class LruCache {
     }
 };
 
-using ImplementationsCache = cldnn::LruCache<size_t, std::shared_ptr<primitive_impl>>;
 using KernelsCache = cldnn::LruCache<size_t, cldnn::kernel::ptr>;
+
+template<typename Key, typename Value>
+class LruCacheThreadSafe {
+public:
+    using LRUCacheType = cldnn::LruCache<Key, Value>;
+    using FuncRemoveItem = std::function<void(Key)>;
+
+    explicit LruCacheThreadSafe(size_t caps) {
+        _impl = cldnn::make_unique<LRUCacheType>(caps);
+    }
+
+    bool add(const Key& key, const Value& value) {
+        std::lock_guard<std::mutex> lock(_mutex);
+        auto popped_lru_key = _impl->get_lru_key();
```

Comment:
We cannot predict whehter the custom function _remove_popped_items needs key or value. 
How about get both (pair of key, value) and let the custom function to use any of them?

Reply:
Done.